hare/algorithm.go

```diff
@@ -554,18 +548,47 @@ func (proc *consensusProcess) onRoundEnd(ctx context.Context) {
 	}
 }
 
-// advances the state to the next round.
-func (proc *consensusProcess) advanceToNextRound(ctx context.Context) {
+// advanceToNextRound advances the state to the next round.
+// Returns k, the new round number.
+// Will return an error if unable to advance to next round.
+// The consensuProcess should be stopped if an error is returned.
+func (proc *consensusProcess) advanceToNextRound(ctx context.Context) (uint32, error) {
+	k := proc.getK()
+	// Edge cases:
+	// 1. Preround -> just continue
+	if k == preRound {
+		k = proc.addToK(1)
+		return k, nil
+	}
+	// 2. Already in certify round -> error
+	if k == certifyRound {
+		proc.With().Error("consensus process trying to advance past certify round",
+			log.Uint32("current_k", k),
+			proc.instanceID)
+		return k, errors.New("already at certify round")
+	}
+	// 3. Notify round AND consensus has been reached -> goto certify round
 	if proc.completed && proc.currentRound() == notifyRound {
 		proc.setK(certifyRound)
-	} else {
-		k := proc.addToK(1)
-		if k >= 4 && k%4 == 0 {
-			proc.WithContext(ctx).Event().Warning("starting new iteration",
-				log.Uint32("current_k", k),
-				proc.instanceID)
-		}
+		return k, nil
+	}
+	// 4. Not Pre round and not Certify round, but k is greater than limit -> error
+	if k >= uint32(proc.cfg.LimitIterations)*RoundsPerIteration {
```

Comment:
the increment of k needs to happen before this check. otherwise will will run an extra status round.

Reply:
not sure I see what you mean here.

Reply:
say k is 39 here, (in testnet `LimitIterations` is 10). it will pass this check and go ahead to run round 40 and exit when you try to advance to round 41. the old code will skip round 40 and exit, which is the correct behavior.

Reply:
Yes, I see where my brain cell failed me. Another brain cell of mine didn't want to actually increment until we were ready to exit the function due to potential side effects on termination and logging confusion. So as a compromise (and better solution) I will _simulate the increment_ 🙃